[Sofa.Type] Clean Quat

damienmarchal opened the pull request:
- moves code from header to .inl
- use typedef to express the inner types instead of specifying them all the time
  The inner types are: Quat, Vec3, Mat3x3, Real
- add Quaternion to forward declaration file.
- remove the use of the implicit automatically converting functions(BREAKING) 
eg: template<Real2> void toMatrix(Matrix<3,3,Real2>){}
As there is no more hidden type conversion the caller's have to handle themselves
the type conversion before calling instead of relying on the inner-conversion. This is
good as it propagate to the user-level that there is type conversion involved instead of
hidden that fact. The consequence is that it is not possible to do stuff like that anymore
Quat<float>::toMatrix(Mat<3,3,double)>

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

damienmarchal committed:
SofaMiscEngine: Fix DisplacementMatrixEngine that was using the generic Rigid3Coord but forcing Matrix4x4f)

damienmarchal committed:
[Sofa.Type] Add Vec forward declaration in type/fwd.h

[Sofa.Type] Remove declaration of alias from Vec.h as they are in type/fwd.h

damienmarchal committed:
[Sofa.Type] Refactoring of Quat.

damienmarchal committed:
[Sofa.Type] Add forward declaration for quaternion in type/fwd.h

[Sofa.Type] Add missing Sofa.Type in Quat.h

damienmarchal committed:
FIXUP windows (as usual).

damienmarchal committed:
[Sofa.Type] Add SOFA_BUILD_SOFA_TYPE in CMakeLists.txt

Signed-off-by: Damien Marchal <damien.marchal@univ-lille1.fr>

damienmarchal committed:
[Sofa.Type] Add missing SOFA_TYPE_API in Quat

damienmarchal committed:
[Sofa.Type] FIX Quat warning about size conversion.

damienmarchal committed:
[SofaDefaultType] Add Sofa.Type as one of the  target_link_libraries.

damienmarchal committed:
Merge remote-tracking branch 'cristal/pr-quaternion-clean-step1' into pr-quaternion-clean-step1

damienmarchal committed:
[Sofa.Type] Restore thresold behavior and fix mistake in writeOpenGlMatrix

damienmarchal committed:
[SofaRigid] Replace RigidMapping<Rigid3Types,Vec3dTypes> to use Vec3Types

Because mixing types is not consistant

damienmarchal committed:
[SofaBaseVisual_test] Replace the use of deprecated sofa::loader::Material()

Material is now in sofa::helper::types()

damienmarchal committed:
Merge remote-tracking branch 'upstream/master' into pr-quaternion-clean-step1

damienmarchal commented:
[ci-build][with-scene-tests][with-regression-tests]

alxbilger commented on SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.h:
```diff
@@ -92,74 +94,36 @@ class SOFA_TYPE_API Quat
     }
 
-    void fromFrame(type::Vec<3,Real>& x, type::Vec<3,Real>&y, type::Vec<3,Real>&z);
+    void fromFrame(Vec3& x, Vec3&y, Vec3&z);
```
```suggestion
    void fromFrame(const Vec3& x, const Vec3&y, const Vec3&z);
```

alxbilger commented on SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.inl:
```diff
@@ -280,32 +269,29 @@ void Quat<Real>::normalize()
 
 template<class Real>
-void Quat<Real>::fromFrame(type::Vec<3,Real>& x, type::Vec<3,Real>&y, type::Vec<3,Real>&z)
+void Quat<Real>::fromFrame(Vec3& x, Vec3&y, Vec3&z)
```
```suggestion
void Quat<Real>::fromFrame(const Vec3& x, const Vec3&y, const Vec3&z)
```

alxbilger commented on SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.h:
```diff
@@ -92,74 +94,36 @@ class SOFA_TYPE_API Quat
-    Quat quatVectMult(const type::Vec<3,Real>& vect);
-
-    Quat vectQuatMult(const type::Vec<3,Real>& vect);
+    auto quatVectMult(const Vec3& vect) -> Quat;
```
```suggestion
    auto quatVectMult(const Vec3& vect) const -> Quat;
```

alxbilger commented on SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.h:
```diff
@@ -92,74 +94,36 @@ class SOFA_TYPE_API Quat
-
-    Quat vectQuatMult(const type::Vec<3,Real>& vect);
+    auto quatVectMult(const Vec3& vect) -> Quat;
+    auto vectQuatMult(const Vec3& vect) -> Quat;
```
```suggestion
    auto vectQuatMult(const Vec3& vect) const -> Quat;
```

alxbilger commented on SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.inl:
```diff
@@ -189,36 +182,32 @@ void Quat<Real>::operator/=(const Real& r)
 
 template<class Real>
-Quat<Real> Quat<Real>::quatVectMult(const type::Vec<3,Real>& vect)
+auto Quat<Real>::quatVectMult(const Vec3& vect) -> Quat
```
```suggestion
auto Quat<Real>::quatVectMult(const Vec3& vect) const -> Quat
```

alxbilger commented on SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.inl:
```diff
@@ -189,36 +182,32 @@ void Quat<Real>::operator/=(const Real& r)
 
 template<class Real>
-Quat<Real> Quat<Real>::vectQuatMult(const type::Vec<3,Real>& vect)
+auto Quat<Real>::vectQuatMult(const Vec3& vect) -> Quat
```
```suggestion
auto Quat<Real>::vectQuatMult(const Vec3& vect) const -> Quat
```

alxbilger commented on SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.h:
```diff
@@ -228,154 +173,40 @@ class SOFA_TYPE_API Quat
-        return quatResult;
-    }
-
+    static auto createQuaterFromEuler( Vec3 v, EulerOrder order = EulerOrder::ZYX) -> Quat;
```
```suggestion
    static auto createQuaterFromEuler(const Vec3& v, EulerOrder order = EulerOrder::ZYX) -> Quat;
```

alxbilger commented on SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.inl:
```diff
@@ -801,4 +729,206 @@ void Quat<Real>::operator*=(const Quat<Real>& q1)
+
+
+template<class Real>
+auto Quat<Real>::createQuaterFromEuler( Vec3 v, EulerOrder order) -> Quat
```
```suggestion
auto Quat<Real>::createQuaterFromEuler(const Vec3& v, EulerOrder order) -> Quat
```

alxbilger commented on SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.h:
```diff
@@ -228,154 +173,40 @@ class SOFA_TYPE_API Quat
-        Quat q = b.inverse() * a;
-        return q;
-    }
+    auto quatDiff( Quat a, const Quat& b) -> Quat;
```
```suggestion
    static auto quatDiff( Quat a, const Quat& b) -> Quat;
```

alxbilger commented on SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.h:
```diff
@@ -228,154 +173,40 @@ class SOFA_TYPE_API Quat
-                                                        // this is done to keep the old behavior (before the
-                                                        // correction of the toEulerVector function).
-    }
+    auto angularDisplacement( Quat a, const Quat& b) -> Vec3;
```
```suggestion
    static auto angularDisplacement( const Quat& a, const Quat& b) -> Vec3;
```

alxbilger commented on SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.inl:
```diff
@@ -801,4 +729,206 @@ void Quat<Real>::operator*=(const Quat<Real>& q1)
+
+/// Return the eulerian vector resulting of the movement between 2 quaternions
+template<class Real>
+auto Quat<Real>::angularDisplacement(Quat a, const Quat& b) -> Vec3
```
```suggestion
auto Quat<Real>::angularDisplacement(const Quat& a, const Quat& b) -> Vec3
```

alxbilger commented on SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.h:
```diff
@@ -228,154 +173,40 @@ class SOFA_TYPE_API Quat
-    Quat<Real> slerp(Quat<Real> &q1, Real t);
-    Quat<Real> slerp2(Quat<Real> &q1, Real t);
+
+    auto slerp(Quat &q1, Real t) -> Quat;
```
```suggestion
    auto slerp(const Quat &q1, Real t) const -> Quat;
```

alxbilger commented on SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.h:
```diff
@@ -228,154 +173,40 @@ class SOFA_TYPE_API Quat
-    Quat<Real> slerp2(Quat<Real> &q1, Real t);
+
+    auto slerp(Quat &q1, Real t) -> Quat;
+    auto slerp2(Quat &q1, Real t) -> Quat;
```
```suggestion
    auto slerp2(const Quat &q1, Real t) const-> Quat;
```

alxbilger commented on SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.inl:
```diff
@@ -655,10 +591,10 @@ Quat<Real> Quat<Real>::slerp(Quat<Real> &q1, Real t)
 // Given an axis and angle, compute quaternion.
 template<class Real>
-Quat<Real> Quat<Real>::slerp2(Quat<Real> &q1, Real t)
+auto Quat<Real>::slerp2(Quat &q1, Real t) -> Quat
```
```suggestion
auto Quat<Real>::slerp2(const Quat &q1, Real t) const -> Quat
```

alxbilger commented on SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.inl:
```diff
@@ -620,26 +564,18 @@ void Quat<Real>::slerp(const Quat& a, const Quat& b, Real t, bool allowFlip)
-
 template<class Real>
-Quat<Real> Quat<Real>::slerp(Quat<Real> &q1, Real t)
+auto Quat<Real>::slerp(Quat &q1, Real t) -> Quat
```
```suggestion
auto Quat<Real>::slerp(const Quat &q1, Real t) const -> Quat
```

alxbilger reviewed (commented):
Minor suggestions related to constness and references

damienmarchal committed:
Apply suggestions from code review

Co-authored-by: alxbilger <alxbilger@users.noreply.github.com>

damienmarchal commented:
Thanks for the suggestion @alxbilger, 

Did you use a tool to make the suggestion on github or did you do that manually ? It is the first time I try the "commit in batch" feature from git. In general we are more often making PR from a suggestion-branch to the PR branch.

alxbilger commented:
> Thanks for the suggestion @alxbilger,
> 
> Did you use a tool to make the suggestion or did you do that manually ? It is the first time I try the "commit in batch" feature from git. In general we are more often making PR from a suggestion-branch to the PR branch.

I used GitHub only. It's also the first time I see that we can directly suggest things in a PR review. GitHub notified me about this new feature when I tried to comment a line the first time. I am not sure it's a good practice, because I did not compile my suggestions at all.

guparan commented on SofaKernel/modules/Sofa.Type/CMakeLists.txt:
```diff
@@ -48,6 +48,7 @@ sofa_find_package(Sofa.Config REQUIRED)
 add_library(${PROJECT_NAME} SHARED ${HEADER_FILES} ${SOURCE_FILES})
 
 target_link_libraries(${PROJECT_NAME} PUBLIC Sofa.Config)
+target_compile_definitions(${PROJECT_NAME} PRIVATE "-D SOFA_BUILD_SOFA_TYPE")
```
```suggestion
target_link_libraries(${PROJECT_NAME} PUBLIC Sofa.Config)
```

guparan commented on SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.inl:
```diff
@@ -801,4 +729,206 @@ void Quat<Real>::operator*=(const Quat<Real>& q1)
+    m[2][0] = (2 * (_q[2] * _q[0] - _q[1] * _q[3]));
+    m[2][1] = (2 * (_q[1] * _q[2] + _q[0] * _q[3]));
+    m[2][2] = (1 - 2 * (_q[1] * _q[1] + _q[0] * _q[0]));
+}
```
@Younesssss remark: why m[3] is not handled here?

damienmarchal committed:
[Sofa.Type] Remove -D SOFA_BUILD_WITH_TYPE in CMakeLists.txt

Not needed because it is already done by the sofa_xxx macro.

damienmarchal committed:
[Sofa.Config] Add SOFA_ATTRIBUTE_CLEANING__QUAT_API to handle deprecation

damienmarchal committed:
[Sofa.Core] Add toHomogenousMatrix(mat4x4) and deprecate toMatrix(mat4x4)

damienmarchal committed:
[Registration] Replace Mat4x4 by Mat3,3 in RegistrationExporter

damienmarchal committed:
Apply suggestions from code review

Co-authored-by: alxbilger <alxbilger@users.noreply.github.com>

damienmarchal committed:
Merge branch 'master' into pr-quaternion-clean-step1

damienmarchal committed:
Merge remote-tracking branch 'upstream/master' into pr-quaternion-clean-step1

damienmarchal committed:
[Sofa.Type] Fix toHomogenousMatrix code

damienmarchal commented:
[ci-build][with-scene-tests][with-regression-tests]

damienmarchal committed:
SofaSimulationCore] Makes extern template the NodeSequence object

Next step will be to remove the use of Link/MultiLink for this code and use
a "normale" class.

damienmarchal committed:
[SofaBoundaryCondition] Remove #include<BaseMeshtTopology.h> in PointConstraint

damienmarchal committed:
[SofaDefaultType] Remove #include<sstream and Messaging.h> in DataTypeInfo.h

damienmarchal committed:
[Sofa.Type] Remove #include<algorithm> from fixed_array.h

damienmarchal committed:
[Sofa.Type] Move #include<cmath> out of fixed_array.h

damienmarchal committed:
Merge remote-tracking branch 'upstream/master' into pr-quaternion-clean-step1

damienmarchal committed:
Merge remote-tracking branch 'cristal/pr-quaternion-clean-step1' into pr-quaternion-clean-step1

# Conflicts:
#	SofaKernel/modules/Sofa.Type/src/sofa/type/Quat.h

damienmarchal commented:
There is PR in SofaPython3 to keep synchronized with this PR. 
https://github.com/sofa-framework/SofaPython3/pull/133

damienmarchal committed:
Merge remote-tracking branch 'upstream/master' into pr-quaternion-clean-step1

# Conflicts:
#	SofaKernel/modules/SofaHelper/src/sofa/helper/GenerateRigid.h

damienmarchal commented:
[ci-build][with-scene-tests][with-regression-tests]

damienmarchal committed:
Merge branch 'master' into pr-quaternion-clean-step1

damienmarchal committed:
Merge remote-tracking branch 'cristal/pr-quaternion-clean-step1' into pr-quaternion-clean-step1

damienmarchal committed:
To REVERT (sychronization with SofaPython PR #133)
